Individual.cpp

```diff
@@ -25,10 +25,55 @@ Individual::Individual(const Graph &graph) {
     generate_neighbour_vector(graph);
 }
 
+// create an offspring from two parents i.e. perform crossover between two parents
+// here, a binary mask of length equal to the number of nodes is randomly
+// created, and an offspring is generated by selecting from the
+// first parent the genes where the mask is 0, and from the second
+// parent the genes where the mask is 1.
+Individual::Individual(const Individual &individual1, const Individual &individual2) {
+    int no_chromosomes = individual1.size();
+
+    for (int i = 0; i < no_chromosomes; i++) {
+        int bit = rand() % 2;
+
+        if (bit == 0) {
+            chosen_neighbour_vector_.push_back(individual1[i]);
+        } else {
+            chosen_neighbour_vector_.push_back(individual2[i]);
+        }
+    }
+}
+
 std::ostream &operator<<(std::ostream &os, const Individual &individual) {
     for (int i : individual.chosen_neighbour_vector_) {
         os << i << " ";
     }
 
     return os;
 }
+
+int Individual::size() const {
+    return chosen_neighbour_vector_.size();
+}
+
+int Individual::operator[](const int index) const {
+    return chosen_neighbour_vector_[index];
+}
+
+// mutate each chromosome of the individual with a given probability;
+// for each chromosome, a random new neighbour is selected
+// and assigned as the new allele value of that chromosome
+void Individual::mutate(const Graph &graph, int mutation_probability) {
+    for (int i = 0; i < chosen_neighbour_vector_.size(); i++) {
+        double current_mutation_probability = ((double) rand() / (RAND_MAX));
+
+        if (current_mutation_probability < mutation_probability) {
```

Comment:
mutation_probability is int while current_mutation_probability is double in the interval [0, 1]